source/mainnet/smart-contracts/guides/factory-pattern.rst

```diff
@@ -0,0 +1,392 @@
+
+The sequence diagram below illustrates an idealised factory pattern.
+The user invokes the ``produce`` endpoint on a ``Factory`` contract.
+The factory constructs a new instance of the ``product`` smart contract, invoking the
+``init_product`` constructor.
+Finally, a reference to the new product instance is returned.
+
+.. image:: images/ideal-factory.svg
+    :alt: sequence diagram showing the idealised interaction for the factory pattern
+
+
+Unfortunately, implementing a factory pattern on Concordium is complicated by the fact that one
+smart contract instance cannot create other smart contract instances programmatically.
+On Concordium, every smart contract instance is created by a top-level transaction. To achieve
+something resembling the factory pattern, we have to separately create the new product instances
+in an uninitialized state, and then have the factory initialize them.
+
+The sequence diagram below illustrates how the factory pattern is realised on Concordium.
+First of all, the user constructs a new instance of the ``product`` smart contract, invoking the
+``init_product`` constructor.
+This creates the ``Product`` contract in a special uninitialized state.
+The user then invokes the ``produce`` endpoint on the ``Factory`` contract, passing in a reference
+to the ``Product`` contract.
+The ``Factory`` contract in turn calls ``initialize`` on the ``Product`` contract, resulting in
+the ``Product`` being fully initialized.
+
+.. image:: images/concordium-factory.svg
+    :alt: sequence diagram showing how the factory pattern might be realised on Concordium
+
+This process is significantly more complex than the idealised factory pattern we started with.
+In particular, the user is required to sign two separate transactions, with the second one depending
+on the result of the first one. If this two-step process is presented to the end user, it is likely
+to cause confusion. Additionally, two-step process means that the smart contracts have to be robust
+against misuse. For instance, the factory has to deal with the possibility that ``produce`` is
+called with a reference to the wrong type of product contract.
+
+Alternatives to the factory pattern
+===================================
+
+Since there are a number of disadvantages to using the factory pattern on Concordium, it is
+important to consider whether a better alternative is available. One of the following alternatives
+may be suitable, depending on the application.
+
+Construct the product directly
+------------------------------
+
+If the factory does not maintain an on-going relationship with the product,
+then it is generally possible to simply initialize the product entirely in
+the constructor (``init_product``), removing the separate ``initialize`` operation entirely.
+In general, if the factory contract does not need to update its state
+after initializing the product, then it serves no purpose. It would be
```

Comment:
If I understand correctly, you want to state that only if the `factory` stores some data in its state in the same function call when `producing` a new `product` then it has a purpose. 

Since the factory can also update its state before `initializing` the product which is the recommended pattern `check-effect-interact`, I would rephrase this to:

```suggestion
In general, if the factory contract does not need to update its state
when initializing the product, then it serves no purpose. It would be
```